compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -52,6 +52,35 @@
 //! _a = *_b // _b is &Freeze
 //! _c = *_b // replaced by _c = _a
 //! ```
+//!
+//! # Determinism of constant propagation
+//!
+//! When registering a new `Value`, we attempt to opportunistically evaluate it as a constant.
+//! The evaluated form is inserted in `evaluated` as an `OpTy` or `None` if evaluation failed.
+//!
+//! The difficulty is non-deterministic evaluation of MIR constants. Some `Const` can have
+//! different runtime values each time they are evaluated. This is the case with
+//! `Const::Slice` which have a new pointer each time they are evaluated, and constants that
+//! contain a fn pointer (`AllocId` pointing to a `GlobalAlloc::Function`) pointing to a different
+//! symbol in each codegen unit.
+//!
+//! Meanwhile, we want to be able to read indirect constants. For instance:
+//! ```
+//! static A: &'static &'static u8 = &&63;
+//! fn foo() -> u8 {
+//!     **A // We want to replace by 63.
+//! }
+//! fn bar() -> u8 {
+//!     b"abc"[1] // We want to replace by 'b'.
+//! }
+//! ```
+//!
+//! The `Value::Constant` variant stores a possibly unevaluated constant. Evaluating that constant
+//! may be non-deterministic. When that happens, we assign a disambiguator to ensure that we do not
+//! merge the constants. See `duplicate_slice` test in `gvn.rs`.
+//!
+//! Second, when writing constants in MIR, we do not write `Const::Slice` or `Const`
+//! that contain `AllocId`s.
```

Comment:
This is very conservative, right? Almost all AllocId are fine. Only function pointers are not.

Reply:
Yes, this is conservative. I don't think an `AllocId` which recursively points to a function pointer is fine.

Reply:
I'm increasingly feeling like we're going down the completely wrong rabbit hole with this fn ptr stuff. What even is the problematic example here? The semantic idea is that Fn-item-to-fn-ptr coercion is non-deterministic. But when you have an AllocId, that's already after that coercion.

Reply:
The `AllocId` of a function definition (which is what function pointers use) points to an `Instance` rather than a specific instantiated version of the function. Only during codegen is it possible to refer to a specific instantiated version of the function as which one you get depends on inside which codegen unit the current mono item gets codegened.

Reply:
Yes.

But how can GVN ever change the set of possible outcomes here?

Reply:
If we have a `static A = &&(|| {} as fn());` for instance. GVN is able to compute some pointer value for `*A`, but we make need sure not to write it into MIR.
If we were to write the `AllocId` for `*A` in MIR, the codegen-unit will create a LLVM constant to back that AllocId, but containing a pointer to the wrong instance of the closure.

Reply:
`*A` is still a reference to the closure, do you mean `**A`?

I think the way to prevent that is when assigning a number to a constant, we have to check the provenance it contains and if there's any function (or vtable) there we bail out.

Constants with such provenance in them are really more like functions that will eventually be given the address of that function (or vtable) *in the then-current codegen unit* and resolve to that.

Reply:
> `*A` is still a reference to the closure, do you mean `**A`?
> 

I really mean `*A`, to discuss the reference to the fn ptr.

> I think the way to prevent that is when assigning a number to a constant, we have to check the provenance it contains and if there's any function (or vtable) there we bail out.
> 
> Constants with such provenance in them are really more like functions that will eventually be given the address of that function (or vtable) _in the then-current codegen unit_ and resolve to that.

That would be needlessly restrictive. The issue happens when we cross CGU boundary. Inside a single MIR body, an fn pointer is equal to itself, so we can number it.

Reply:
Hm, yes that makes sense if the "they really are functions" model since they will all be given the same argument then.

But I don't understand your example, then. `*A` is just a harmless piece of memory?

Or are you concerned about https://github.com/rust-lang/rust/issues/79738? That's a bug. If `A` is a static then `*A` should be the exact same thing everywhere. We shouldn't design GVN around a bug.

With `const A = &&(|| {} as fn());`, the situation is a bit different, but then `A` is an unevaluated const so we already treat each of its uses separately with your latest patch.

Reply:
I did not know that #79738 was a bug. But for now, it's a relatively hard to trigger. In my earlier attempts around gvn and constprop, neglecting that bug broke the std::fmt, that's why I'm being extra cautious on that side. 
If we manage to resolve that bug, many restrictions around putting AllocId in constants produced by optimisations won't be needed any more.

Reply:
>  neglecting that bug broke the std::fmt,

Oh, ouch. Do you know which fn-ptr-containing-static caused that?

> If we manage to resolve that bug, many restrictions around putting AllocId in constants produced by optimisations won't be needed any more.

Ah, great! Okay so those places should be clearly labeled as "working around #79738" then.

Reply:
https://github.com/rust-lang/rust/blob/4af886f8ab94543caad689ee6bf6a93fa8bd4a98/library/core/src/fmt/rt.rs#L197-L216

Reply:
But that's not a static-containing-nested-allocations. That has the fn ptr directly in the static so even with that bug we shouldn't be duplicating anything.